rust/src/nasl/builtin/ssh/libssh/mod.rs

```diff
@@ -0,0 +1,152 @@
+// SPDX-FileCopyrightText: 2023 Greenbone AG
+//
+// SPDX-License-Identifier: GPL-2.0-or-later WITH x11vnc-openssl-exception
+
+//! Defines NASL ssh and sftp functions
+
+mod channel;
+mod session;
+
+use std::net::{IpAddr, UdpSocket};
+use std::{os::fd::AsRawFd, time::Duration};
+
+use libssh_rs::{LogLevel, SshOption};
+use russh::cipher;
+use russh_keys::key;
+use tokio::sync::{Mutex, MutexGuard};
+use tracing::debug;
+
+use super::error::Result;
+use super::Ssh;
+
+pub use libssh_rs::AuthMethods;
+pub use session::SshSession;
+
+pub type SessionId = i32;
+pub type Socket = std::os::raw::c_int;
+
+pub fn get_log_level() -> LogLevel {
+    let verbose = std::env::var("OPENVAS_LIBSSH_DEBUG")
+        .map(|x| x.parse::<i32>().unwrap_or_default())
+        .unwrap_or(0);
+
+    match verbose {
+        0 => LogLevel::NoLogging,
+        1 => LogLevel::Warning,
+        2 => LogLevel::Protocol,
+        3 => LogLevel::Packet,
+        _ => LogLevel::Functions,
+    }
+}
+
+fn to_comma_separated_string<T: AsRef<str>>(items: &[T]) -> Option<String> {
+    if items.is_empty() {
+        None
+    } else {
+        Some(
+            items
+                .iter()
+                .map(|name| name.as_ref().to_string())
+                .collect::<Vec<_>>()
+                .join(","),
+        )
+    }
+}
+
+type BorrowedSession<'a> = MutexGuard<'a, SshSession>;
+
+impl Ssh {
+    pub async fn find_id<'a>(
+        &'a self,
+        f: impl for<'b> Fn(&BorrowedSession<'b>) -> bool,
+    ) -> Result<Option<SessionId>> {
+        for id in self.ids() {
+            let session = self.get_by_id(*id).await?;
```

Comment:
q: under which criteria does get_by_id fail? I'm asking because it is hard to tell if the result of that would be always Err when ids are available or not.

Reply:
`get_by_id` only fails if the ID does not exist (the locking could throw an error with std mutexes but not with tokio async mutexes).